packages/apriltag/src/apriltag_detector_node.py

```diff
@@ -0,0 +1,185 @@
+        # pack detections into a message
+        msg = AprilTagDetectionArray()
+        detection_time = rospy.Time.now()
+        msg.header.stamp = detection_time
+        msg.header.frame_id = self._camera_frame
+        for tag in tags:
+            # turn rotation matrix into quaternion
+            q = _matrix_to_quaternion(tag.pose_R)
+            p = tag.pose_t.T[0]
+            # create single tag detection object
+            detection = AprilTagDetection(
+                transform=Transform(
+                    translation=Vector3(
+                        x=p[0],
+                        y=p[1],
+                        z=p[2]
+                    ),
+                    rotation=Quaternion(
+                        x=q[0],
+                        y=q[1],
+                        z=q[2],
+                        w=q[3]
+                    )
+                ),
+                tag_id=tag.tag_id,
+                tag_family=tag.tag_family,
+                hamming=tag.hamming,
+                decision_margin=tag.decision_margin,
+                homography=tag.homography.flatten().astype(np.float32).tolist(),
+                center=tag.center.tolist(),
+                corners=tag.corners.flatten().tolist(),
+                pose_error=tag.pose_err
+            )
+            # add detection to array
+            msg.detections.append(detection)
+            # publish tf
+            self._tf_bcaster.sendTransform(
+                p.tolist(),
+                q.tolist(),
+                detection_time,
+                '/tag{:s}'.format(str(tag.tag_id)),
+                self._camera_frame
+            )
+        # publish detections
+        self._tag_pub.publish(msg)
+        # update healthy frequency metadata
+        self._tag_pub.set_healthy_freq(self._img_sub.get_frequency())
+        self._img_pub.set_healthy_freq(self._img_sub.get_frequency())
+
+    def _publish_detections_image(self, img, tags):
+        # get a color buffer from the BW image
+        color_img = cv2.cvtColor(img, cv2.COLOR_GRAY2RGB)
+        # draw each tag
+        for tag in tags:
+            for idx in range(len(tag.corners)):
+                cv2.line(
+                    color_img,
+                    tuple(tag.corners[idx - 1, :].astype(int)),
+                    tuple(tag.corners[idx, :].astype(int)),
+                    (0, 255, 0)
+                )
+            # draw the tag ID
+            cv2.putText(
+                color_img,
+                str(tag.tag_id),
+                org=(tag.corners[0, 0].astype(int) + 10, tag.corners[0, 1].astype(int) + 10),
+                fontFace=cv2.FONT_HERSHEY_SIMPLEX,
+                fontScale=0.8,
+                color=(0, 0, 255)
+            )
+        # pack image into a message
+        img_msg = self._bridge.cv2_to_compressed_imgmsg(color_img)
+        # ---
+        self._img_pub.publish(img_msg)
+        self._img_pub_busy = False
+
+    def _cinfo_cb(self, data):
```

Comment:
we should not shutdown the subscriber. If the camera resolution is changed, we need to update the camera parameters and re-initalize the Detector